File: src/main/java/io/kuzzle/sdk/core/Room.java
package io.kuzzle.sdk.core;

import android.support.annotation.NonNull;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.Timer;
import java.util.TimerTask;
import java.util.UUID;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;

import io.kuzzle.sdk.enums.Event;
import io.kuzzle.sdk.enums.Scope;
import io.kuzzle.sdk.enums.State;
import io.kuzzle.sdk.enums.Users;
import io.kuzzle.sdk.listeners.ResponseListener;
import io.kuzzle.sdk.listeners.SubscribeListener;
import io.kuzzle.sdk.listeners.OnQueryDoneListener;
import io.kuzzle.sdk.responses.NotificationResponse;
import io.kuzzle.sdk.state.KuzzleStates;
import io.socket.emitter.Emitter;

/**
 * The type Kuzzle room.
 */
public class Room {

  private String id = UUID.randomUUID().toString();
  /**
   * The Collection.
   */
  protected String collection;
  /**
   * The Data collection.
   */
  protected Collection dataCollection;
  /**
   * The Filters.
   */
  protected JSONObject filters = new JSONObject();
  /**
   * The Headers.
   */
  protected JSONObject headers;
  /**
   * The Metadata.
   */
  protected JSONObject metadata;
  /**
   * The Subscribe to self.
   */
  protected boolean subscribeToSelf;
  /**
   * The Room id.
   */
  protected String roomId;
  /**
   * The Kuzzle.
   */
  protected Kuzzle kuzzle;
  /**
   * The Channel.
   */
  protected String channel;
  /**
   * The Scope.
   */
  protected Scope scope;
  /**
   * The State.
   */
  protected State state;
  /**
   * The Users.
   */
  protected Users users;
  /**
   * The Listener.
   */
  protected ResponseListener<NotificationResponse> listener;

  // Used to avoid subscription renewals to trigger multiple times because of
  // multiple but similar events
  private long lastRenewal = 0;
  private long renewalDelay = 500;

  /**
   * The Subscribing.
   */
// Used to delay method calls when subscription is in progress
  protected boolean subscribing = false;
  private ArrayList<Runnable> queue = new ArrayList<>();
  private SubscribeListener doneListener;

  /**
   * Instantiates a new Kuzzle room.
   *
   * @param kuzzleDataCollection the kuzzle data collection
   */
  public Room(@NonNull final Collection kuzzleDataCollection) {
    this(kuzzleDataCollection, null);
  }

  /**
   * This object is the result of a subscription request, allowing to manipulate the subscription itself.
   * In Kuzzle, you don't exactly subscribe to a room or a topic but, instead, you subscribe to documents.
   * What it means is that, to subscribe, you provide to Kuzzle a set of matching filters.
   * Once you have subscribed, if a pub/sub message is published matching your filters, or if a matching stored
   * document change (because it is created, updated or deleted), then you'll receive a notification about it.
   *
   * @param kuzzleDataCollection the kuzzle data collection
   * @param options              the options
   */
  public Room(@NonNull final Collection kuzzleDataCollection, final RoomOptions options) {
    RoomOptions opts = options != null ? options : new RoomOptions();

    if (kuzzleDataCollection == null) {
      throw new IllegalArgumentException("Room: missing dataCollection");
    }
    kuzzleDataCollection.getKuzzle().isValid();

    this.dataCollection = kuzzleDataCollection;
    this.kuzzle = kuzzleDataCollection.getKuzzle();
    this.collection = kuzzleDataCollection.getCollection();

    try {
      this.headers = new JSONObject(kuzzleDataCollection.getHeaders().toString());
    }
    catch (JSONException e) {
      throw new RuntimeException(e);
    }

    this.subscribeToSelf = opts.isSubscribeToSelf();
    this.metadata = opts.getMetadata();
    this.scope = opts.getScope();
    this.state = opts.getState();
    this.users = opts.getUsers();
  }

  /**
   * Returns the number of other subscriptions on that room.
   *
   * @param listener the listener
   */
  public void count(@NonNull final ResponseListener<Integer> listener) {
    if (listener == null) {
      throw new IllegalArgumentException("Room.count: a callback listener is required");
    }

    // Delays this call until after the subscription is finished
    if (!this.isReady()) {
      this.queue.add(new Runnable() {
        @Override
        public void run() {
          Room.this.count(listener);
        }
      });
      return;
    }

    if (this.roomId == null) {
      throw new IllegalStateException("Room.count: cannot count subscriptions on an inactive room");
    }

    try {
      JSONObject data = new JSONObject().put("body", new JSONObject().put("roomId", this.roomId));
      this.kuzzle.addHeaders(data, this.headers);

      this.kuzzle.query(this.dataCollection.makeQueryArgs("realtime", "count"), data, new OnQueryDoneListener() {
        @Override
        public void onSuccess(JSONObject response) {
          try {
            listener.onSuccess(response.getJSONObject("result").getInt("count"));
          } catch (JSONException e) {
            throw new RuntimeException(e);
          }
        }

        @Override
        public void onError(JSONObject error) {
          listener.onError(error);
        }
      });
    } catch (JSONException e) {
      throw new RuntimeException(e);
    }
  }

  /**
   * Call after renew.
   *
   * @param args the args
   */
  protected void callAfterRenew(final Object args) {
    if (args == null) {
      throw new IllegalArgumentException("Room.renew: response required");
    }

    try {
      if (!((JSONObject) args).isNull("error")) {
        listener.onError((JSONObject) args);
      }
      else {
        String key = ((JSONObject) args).getString("requestId");

        if (((JSONObject) args).getString("action").equals("jwtTokenExpired")) {
          Room.this.kuzzle.jwtToken = null;
          Room.this.kuzzle.emitEvent(Event.jwtTokenExpired);
        }

        if (Room.this.kuzzle.getRequestHistory().containsKey(key)) {
          if (Room.this.subscribeToSelf) {
            listener.onSuccess(new NotificationResponse(kuzzle, (JSONObject) args));
          }
          Room.this.kuzzle.getRequestHistory().remove(key);
        } else {
          listener.onSuccess(new NotificationResponse(kuzzle, (JSONObject) args));
        }
      }
    } catch (JSONException e) {
      try {
        listener.onError(((JSONObject) args).getJSONObject("error"));
      } catch (JSONException err) {
        throw new RuntimeException(e);
      }
    }
  }

  /**
   * Renew the subscription. Force a resubscription using the same filters if no new ones are provided.
   * Unsubscribes first if this Room was already listening to events.
   *
   * @param listener the listener
   * @return kuzzle room
   */
  public Room renew(@NonNull final ResponseListener<NotificationResponse> listener) {
    return this.renew(null, listener, null);
  }

  /**
   * Renew the subscription. Force a resubscription using the same filters if no new ones are provided.
   * Unsubscribes first if this Room was already listening to events.
   *
   * @param listener the listener
   * @param subscribeResponseListener
   * @return kuzzle room
   */
  public Room renew(@NonNull final ResponseListener<NotificationResponse> listener, final SubscribeListener subscribeResponseListener) {
    return this.renew(null, listener, subscribeResponseListener);
  }

  /**
   * Renew the subscription. Force a resubscription using the same filters if no new ones are provided.
   * Unsubscribes first if this Room was already listening to events.
   *
   * @param filters  the filters
   * @param listener the listener
   * @return kuzzle room
   */
  public Room renew(final JSONObject filters, @NonNull final ResponseListener<NotificationResponse> listener, final SubscribeListener subscribeResponseListener) {
    long now = System.currentTimeMillis();

    if (listener == null) {
      throw new IllegalArgumentException("Room.renew: a callback listener is required");
    }

    // Skip subscription renewal if another one was performed just a moment before
    if (this.lastRenewal > 0 && (now - this.lastRenewal) <= this.renewalDelay) {
      return this;
    }

    if (filters != null) {
      this.filters = filters;
    }

    /*
      If not yet connected, registers itself into the subscriptions list and wait for the
      main Kuzzle object to renew subscriptions once online
     */
    if (this.kuzzle.state != KuzzleStates.CONNECTED) {
      this.listener = listener;
      this.doneListener = subscribeResponseListener;
      this.kuzzle.addPendingSubscription(this.id, this);
      return this;
    }

    if (this.subscribing) {
      this.queue.add(new Runnable() {
        @Override
        public void run() {
          Room.this.renew(filters, listener, subscribeResponseListener);
        }
      });

      return this;
    }

    this.unsubscribe();
    this.roomId = null;
    this.subscribing = true;
    this.listener = listener;
    this.doneListener = subscribeResponseListener;
    this.kuzzle.addPendingSubscription(this.id, this);

    try {
      final Options options = new Options();
      final JSONObject
          subscribeQuery = new JSONObject()
          .put("body", this.filters)
          .put("scope", this.scope.toString().toLowerCase())
          .put("state", this.state.toString().toLowerCase())
          .put("users", this.users.toString().toLowerCase());

      options.setMetadata(this.metadata);
      this.kuzzle.addHeaders(subscribeQuery, this.headers);

      new Thread(new Runnable() {
        @Override
        public void run() {
          try {
            Room.this.kuzzle.query(Room.this.dataCollection.makeQueryArgs("realtime", "subscribe"), subscribeQuery, options, new OnQueryDoneListener() {
              @Override
              public void onSuccess(JSONObject args) {
                try {
                  Room.this.kuzzle.deletePendingSubscription(Room.this.id);
                  Room.this.subscribing = false;
                  Room.this.lastRenewal = System.currentTimeMillis();

                  JSONObject result = args.getJSONObject("result");
                  Room.this.channel = result.getString("channel");
                  Room.this.roomId = result.getString("roomId");
                  if (subscribeResponseListener != null) {
                    subscribeResponseListener.done(null, Room.this);
                  }
                } catch (JSONException e) {
                  throw new RuntimeException(e);
                }

                Room.this.kuzzle.addSubscription(Room.this.roomId, Room.this.id, Room.this);

                Room.this.kuzzle.getSocket().on(Room.this.channel, new Emitter.Listener() {
                  @Override
                  public void call(final Object... args) {
                    callAfterRenew(args[0]);
                  }
                });

                Room.this.dequeue();
              }

              @Override
              public void onError(JSONObject arg) {
                Room.this.subscribing = false;
                Room.this.queue.clear();
                if (subscribeResponseListener != null) {
                  subscribeResponseListener.done(arg, null);
                }
              }
            });
          } catch (JSONException e) {
            throw new RuntimeException(e);
          }
        }
      }).start();

    } catch (JSONException e) {
      throw new RuntimeException(e);
    }
    return this;
  }

  /**
   * Unsubscribes from Kuzzle.
   * Stop listening immediately. If there is no listener left on that room, sends an unsubscribe request to Kuzzle, once
   * pending subscriptions reaches 0, and only if there is still no listener on that room.
   * We wait for pending subscriptions to finish to avoid unsubscribing while another subscription on that room is
   *
   * @return the kuzzle room
   */
  public Room unsubscribe() {
    if (!this.isReady()) {
      this.queue.add(new Runnable() {
        @Override
        public void run() {
          Room.this.unsubscribe();
        }
      });
      return this;
    }

    if (this.roomId == null) {
      return this;
    }

    try {
      final JSONObject data = new JSONObject().put("body", new JSONObject().put("roomId", this.roomId));
      this.kuzzle.addHeaders(data, this.headers);

      this.kuzzle.getSocket().off(Room.this.channel);
      this.kuzzle.deleteSubscription(this.roomId, this.id);

      if (this.kuzzle.getSubscriptions(this.roomId) == null) {
        final String roomId = this.roomId;
        if (this.kuzzle.getPendingSubscriptions().isEmpty()) {
          this.kuzzle.query(this.dataCollection.makeQueryArgs("realtime", "unsubscribe"), data);
        } else {
          final Timer timer = new Timer(UUID.randomUUID().toString());
          unsubscribeTask(timer, roomId, data).run();
        }
      }

      this.roomId = null;
    }
    catch (JSONException e) {
      throw new RuntimeException(e);
    }
    return this;
  }

  /**
   * Unsubscribe task timer task.
   *
   * @param timer  the timer
   * @param roomId the room id
   * @param data   the data
   * @return the timer task
   */
  protected TimerTask unsubscribeTask(final Timer timer, final String roomId, final JSONObject data) {
    return new TimerTask() {
      @Override
      public void run() {
        try {
          if (Room.this.kuzzle.getPendingSubscriptions().isEmpty()) {
            if (Room.this.kuzzle.getSubscriptions(roomId) == null) {
              Room.this.kuzzle.query(Room.this.dataCollection.makeQueryArgs("realtime", "unsubscribe"), data);
            }
          } else {
            timer.schedule(unsubscribeTask(timer, roomId, data), 100);
          }
        } catch (JSONException e) {
          throw new RuntimeException(e);
        }
      }
    };
  }

  /**
   * Gets collection.
   *
   * @return the collection
   */
  public String getCollection() {
    return collection;
  }

  /**
   * Gets filters.
   *
   * @return the filters
   */
  public JSONObject getFilters() {
    return filters;
  }

  /**
   * Sets filters.
   *
   * @param filters the filters
   * @return the filters
   */
  public Room setFilters(final JSONObject filters) {
    this.filters = filters;
    return this;
  }

  /**
   * Gets headers.
   *
   * @return the headers
   */
  public JSONObject getHeaders() {
    return this.headers;
  }

  /**
   * Helper function allowing to set headers while chaining calls.
   * If the replace argument is set to true, replace the current headers with the provided content.
   * Otherwise, it appends the content to the current headers, only replacing already existing values
   *
   * @param content the headers
   * @return the headers
   */
  public Room setHeaders(final JSONObject content) {
    return this.setHeaders(content, false);
  }

  /**
   * Helper function allowing to set headers while chaining calls.
   * If the replace argument is set to true, replace the current headers with the provided content.
   * Otherwise, it appends the content to the current headers, only replacing already existing values
   *
   * @param content - new headers content
   * @param replace - default: false = append the content. If true: replace the current headers with tj
   * @return the headers
   */
  public Room setHeaders(final JSONObject content, final boolean replace) {
    if (this.headers == null) {
      this.headers = new JSONObject();
    }
    if (replace) {
      this.headers = content;
    } else {
      if (content != null) {
        try {
          for (Iterator ite = content.keys(); ite.hasNext(); ) {
            String key = (String) ite.next();
            this.headers.put(key, content.get(key));
          }
        } catch (JSONException e) {
          throw new RuntimeException(e);
        }
      }
    }
    return this;
  }

  /**
   * Gets metadata.
   *
   * @return the metadata
   */
  public JSONObject getMetadata() {
    return metadata;
  }

  /**
   * Sets metadata.
   *
   * @param metadata the metadata
   * @return the metadata
   */
  public Room setMetadata(final JSONObject metadata) {
    this.metadata = metadata;
    return this;
  }

  /**
   * Is subscribe to self boolean.
   *
   * @return the boolean
   */
  public boolean isSubscribeToSelf() {
    return subscribeToSelf;
  }

  /**
   * Sets subscribe to self.
   *
   * @param subscribeToSelf the subscribe to self
   * @return the subscribe to self
   */
  public Room setSubscribeToSelf(final boolean subscribeToSelf) {
    this.subscribeToSelf = subscribeToSelf;
    return this;
  }

  /**
   * Get roomId
   *
   * @return roomId room id
   */
  public String getRoomId() {
    return this.roomId;
  }

  /**
   * Getter for the listener property
   *
   * @return listener
   */
  public ResponseListener<NotificationResponse> getListener() {
    return this.listener;
  }

  public SubscribeListener getSubscribeListener() {
    return doneListener;
  }

  /**
   * Runs all queued methods called while subscription was in progress
   */
  protected void dequeue() {
    if (this.queue.size() > 0) {
      ExecutorService threadPool = Executors.newSingleThreadExecutor();

      for(Runnable r: this.queue) {
        threadPool.execute(r);
      }

      threadPool.shutdown();

      try {
        threadPool.awaitTermination(1, TimeUnit.SECONDS);
      }
      catch (InterruptedException e) {
        // do nothing
      }
      finally {
        this.queue.clear();
      }
    }
  }

  private boolean isReady() {
    return this.kuzzle.state == KuzzleStates.CONNECTED && !this.subscribing;
  }
}

File: src/main/java/io/kuzzle/sdk/core/Room.kt
package io.kuzzle.sdk.core

import org.json.JSONException
import org.json.JSONObject

import java.util.ArrayList
import java.util.Timer
import java.util.TimerTask
import java.util.UUID
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors
import java.util.concurrent.TimeUnit

import io.kuzzle.sdk.enums.Event
import io.kuzzle.sdk.enums.Scope
import io.kuzzle.sdk.enums.State
import io.kuzzle.sdk.enums.Users
import io.kuzzle.sdk.listeners.ResponseListener
import io.kuzzle.sdk.listeners.SubscribeListener
import io.kuzzle.sdk.listeners.OnQueryDoneListener
import io.kuzzle.sdk.responses.NotificationResponse
import io.kuzzle.sdk.state.States
import io.socket.emitter.Emitter

open class Room
/**
 * This object is the result of a subscription request, allowing to manipulate the subscription itself.
 * In Kuzzle, you don't exactly subscribe to a room or a topic but, instead, you subscribe to documents.
 * What it means is that, to subscribe, you provide to Kuzzle a set of matching filters.
 * Once you have subscribed, if a pub/sub message is published matching your filters, or if a matching stored
 * document change (because it is created, updated or deleted), then you'll receive a notification about it.
 *
 * @param kuzzleDataCollection Data collection to link
 * @param options              Subscription options
 */
@JvmOverloads constructor(protected var dataCollection: Collection, options: RoomOptions? = null) {

    private val id = UUID.randomUUID().toString()
    /**
     * Linked data collection name getter
     *
     * @return linked data collection name
     */
    var collection: String
        protected set
    protected var filters = JSONObject()
    protected var headers: JSONObject? = null
    protected var _volatile: JSONObject
    protected var subscribeToSelf: Boolean = false
    /**
     * roomId property getter
     *
     * @return roomId property value
     */
    var roomId: String? = null
        protected set
    protected var kuzzle: Kuzzle
    protected var channel: String
    protected var scope: Scope
    protected var state: State
    protected var users: Users
    /**
     * listener property getter
     *
     * @return listener property value
     */
    var listener: ResponseListener<NotificationResponse>
        protected set

    // Used to avoid subscription renewals to trigger multiple times because of
    // multiple but similar events
    private var lastRenewal: Long = 0
    private val renewalDelay: Long = 500

    // Used to delay method calls when subscription is in progress
    protected var subscribing = false
    private val queue = ArrayList<Runnable>()
    /**
     * subscribeListener property getter
     * @return subscribeListener property value
     */
    var subscribeListener: SubscribeListener? = null
        private set

    private val isReady: Boolean
        get() = this.kuzzle.state == States.CONNECTED && !this.subscribing

    init {
        val opts = options ?: RoomOptions()

        if (dataCollection == null) {
            throw IllegalArgumentException("Room: missing dataCollection")
        }
        dataCollection.kuzzle.isValid()
        this.kuzzle = dataCollection.kuzzle
        this.collection = dataCollection.collection

        try {
            this.headers = JSONObject(dataCollection.getHeaders().toString())
        } catch (e: JSONException) {
            throw RuntimeException(e)
        }

        this.subscribeToSelf = opts.isSubscribeToSelf
        this._volatile = opts.volatile
        this.scope = opts.scope
        this.state = opts.state
        this.users = opts.users
    }

    /**
     * Returns the number of other subscriptions on that room.
     *
     * @param listener Response callback listener
     */
    fun count(listener: ResponseListener<Int>) {
        if (listener == null) {
            throw IllegalArgumentException("Room.count: a callback listener is required")
        }

        // Delays this call until after the subscription is finished
        if (!this.isReady) {
            this.queue.add(Runnable { this@Room.count(listener) })
            return
        }

        if (this.roomId == null) {
            throw IllegalStateException("Room.count: cannot count subscriptions on an inactive room")
        }

        try {
            val data = JSONObject().put("body", JSONObject().put("roomId", this.roomId))
            this.kuzzle.addHeaders(data, this.headers)

            this.kuzzle.query(this.dataCollection.makeQueryArgs("realtime", "count"), data, object : OnQueryDoneListener {
                override fun onSuccess(response: JSONObject) {
                    try {
                        listener.onSuccess(response.getJSONObject("result").getInt("count"))
                    } catch (e: JSONException) {
                        throw RuntimeException(e)
                    }

                }

                override fun onError(error: JSONObject) {
                    listener.onError(error)
                }
            })
        } catch (e: JSONException) {
            throw RuntimeException(e)
        }

    }

    /**
     * Call after renew.
     *
     * @param args the args
     */
    protected open fun callAfterRenew(args: Any?) {
        if (args == null) {
            throw IllegalArgumentException("Room.renew: response required")
        }

        try {
            val requestId = if ((args as JSONObject).has("requestId")) args.getString("requestId") else null

            if (args.getString("type") == "TokenExpired") {
                this@Room.kuzzle.setJwtToken(null)
                this@Room.kuzzle.emitEvent(Event.tokenExpired)
            }

            if (requestId != null && this@Room.kuzzle.requestHistory.containsKey(requestId)) {
                if (this@Room.subscribeToSelf) {
                    listener.onSuccess(NotificationResponse(kuzzle, args as JSONObject?))
                }
                this@Room.kuzzle.requestHistory.remove(requestId)
            } else {
                listener.onSuccess(NotificationResponse(kuzzle, args as JSONObject?))
            }
        } catch (e: JSONException) {
            throw RuntimeException(e)
        }

    }

    /**
     * [.renew]
     */
    fun renew(listener: ResponseListener<NotificationResponse>): Room {
        return this.renew(null, listener, null)
    }

    /**
     * [.renew]
     */
    fun renew(listener: ResponseListener<NotificationResponse>, subscribeResponseListener: SubscribeListener): Room {
        return this.renew(null, listener, subscribeResponseListener)
    }

    /**
     * Renew the subscription. Force a resubscription using the same filters
     * if no new ones are provided.
     * Unsubscribes first if this Room was already listening to events.
     *
     * @param filters  Subscription filters, using Kuzzle DSL
     * @param listener Response callback listener
     * @return this
     */
    fun renew(filters: JSONObject?, listener: ResponseListener<NotificationResponse>, subscribeResponseListener: SubscribeListener?): Room {
        val now = System.currentTimeMillis()

        if (listener == null) {
            throw IllegalArgumentException("Room.renew: a callback listener is required")
        }

        // Skip subscription renewal if another one was performed just a moment before
        if (this.lastRenewal > 0 && now - this.lastRenewal <= this.renewalDelay) {
            return this
        }

        if (filters != null) {
            this.filters = filters
        }

        /*
      If not yet connected, registers itself into the subscriptions list and wait for the
      main Kuzzle object to renew subscriptions once online
     */
        if (this.kuzzle.state != States.CONNECTED) {
            this.listener = listener
            this.subscribeListener = subscribeResponseListener
            this.kuzzle.addPendingSubscription(this.id, this)
            return this
        }

        if (this.subscribing) {
            this.queue.add(Runnable { this@Room.renew(filters, listener, subscribeResponseListener) })

            return this
        }

        this.unsubscribe()
        this.roomId = null
        this.subscribing = true
        this.listener = listener
        this.subscribeListener = subscribeResponseListener
        this.kuzzle.addPendingSubscription(this.id, this)

        try {
            val options = Options()
            val subscribeQuery = JSONObject()
                    .put("body", this.filters)
                    .put("scope", this.scope.toString().toLowerCase())
                    .put("state", this.state.toString().toLowerCase())
                    .put("users", this.users.toString().toLowerCase())

            options.volatile = this._volatile
            this.kuzzle.addHeaders(subscribeQuery, this.headers)

            Thread(Runnable {
                try {
                    this@Room.kuzzle.query(this@Room.dataCollection.makeQueryArgs("realtime", "subscribe"), subscribeQuery, options, object : OnQueryDoneListener {
                        override fun onSuccess(args: JSONObject) {
                            try {
                                this@Room.kuzzle.deletePendingSubscription(this@Room.id)
                                this@Room.subscribing = false
                                this@Room.lastRenewal = System.currentTimeMillis()

                                val result = args.getJSONObject("result")
                                this@Room.channel = result.getString("channel")
                                this@Room.roomId = result.getString("roomId")
                                subscribeResponseListener?.done(null, this@Room)
                            } catch (e: JSONException) {
                                throw RuntimeException(e)
                            }

                            this@Room.kuzzle.addSubscription(this@Room.roomId!!, this@Room.id, this@Room)

                            this@Room.kuzzle.socket!!.on(this@Room.channel) { args -> callAfterRenew(args[0]) }

                            this@Room.dequeue()
                        }

                        override fun onError(arg: JSONObject) {
                            this@Room.subscribing = false
                            this@Room.queue.clear()
                            subscribeResponseListener?.done(arg, null)
                        }
                    })
                } catch (e: JSONException) {
                    throw RuntimeException(e)
                }
            }).start()

        } catch (e: JSONException) {
            throw RuntimeException(e)
        }

        return this
    }

    /**
     * Unsubscribes from Kuzzle.
     * Stop listening immediately. If there is no listener left on that room,
     * sends an unsubscribe request to Kuzzle, once
     * pending subscriptions reaches 0, and only if there is still no listener on that room.
     * We wait for pending subscriptions to finish to avoid unsubscribing while
     * another subscription on that room is
     *
     * @return this
     */
    open fun unsubscribe(): Room {
        if (!this.isReady) {
            this.queue.add(Runnable { this@Room.unsubscribe() })
            return this
        }

        if (this.roomId == null) {
            return this
        }

        try {
            val data = JSONObject().put("body", JSONObject().put("roomId", this.roomId))
            this.kuzzle.addHeaders(data, this.headers)

            this.kuzzle.socket!!.off(this@Room.channel)
            this.kuzzle.deleteSubscription(this.roomId!!, this.id)

            if (this.kuzzle.getSubscriptions(this.roomId) == null) {
                val roomId = this.roomId
                if (this.kuzzle.pendingSubscriptions.isEmpty()) {
                    this.kuzzle.query(this.dataCollection.makeQueryArgs("realtime", "unsubscribe"), data)
                } else {
                    val timer = Timer(UUID.randomUUID().toString())
                    unsubscribeTask(timer, roomId, data).run()
                }
            }

            this.roomId = null
        } catch (e: JSONException) {
            throw RuntimeException(e)
        }

        return this
    }

    /**
     * Unsubscribe task timer task.
     *
     * @param timer  the timer
     * @param roomId the room id
     * @param data   the data
     * @return the timer task
     */
    protected open fun unsubscribeTask(timer: Timer, roomId: String, data: JSONObject): TimerTask {
        return object : TimerTask() {
            override fun run() {
                try {
                    if (this@Room.kuzzle.pendingSubscriptions.isEmpty()) {
                        if (this@Room.kuzzle.getSubscriptions(roomId) == null) {
                            this@Room.kuzzle.query(this@Room.dataCollection.makeQueryArgs("realtime", "unsubscribe"), data)
                        }
                    } else {
                        timer.schedule(unsubscribeTask(timer, roomId, data), 100)
                    }
                } catch (e: JSONException) {
                    throw RuntimeException(e)
                }

            }
        }
    }

    /**
     * Subscription filters getter
     *
     * @return subscription filters
     */
    fun getFilters(): JSONObject {
        return filters
    }

    /**
     * Subscription filters setter.
     * renew must be called for this to take effect
     *
     * @param filters New subscription filters
     * @return this
     */
    fun setFilters(filters: JSONObject): Room {
        this.filters = filters
        return this
    }

    /**
     * headers property getters
     *
     * @return headers value
     */
    fun getHeaders(): JSONObject? {
        return this.headers
    }

    /**
     * [.setHeaders]
     */
    fun setHeaders(content: JSONObject): Room {
        return this.setHeaders(content, false)
    }

    /**
     * Subscription headers setter
     *
     * @param content - new headers content
     * @param replace - default: false = append the content, true = replace
     * @return this
     */
    fun setHeaders(content: JSONObject?, replace: Boolean): Room {
        if (this.headers == null) {
            this.headers = JSONObject()
        }
        if (replace) {
            this.headers = content
        } else {
            if (content != null) {
                try {
                    val ite = content.keys()
                    while (ite.hasNext()) {
                        val key = ite.next() as String
                        this.headers!!.put(key, content.get(key))
                    }
                } catch (e: JSONException) {
                    throw RuntimeException(e)
                }

            }
        }
        return this
    }

    /**
     * Gets volatile data.
     *
     * @return the volatile property
     */
    fun getVolatile(): JSONObject {
        return _volatile
    }

    /**
     * Sets volatile metadata.
     *
     * @param _volatile New volatile data value
     * @return this
     */
    fun setVolatile(_volatile: JSONObject): Room {
        this._volatile = _volatile
        return this
    }

    /**
     * subscribeToSelf property getter
     *
     * @return subscribeToSelf property value
     */
    fun isSubscribeToSelf(): Boolean {
        return subscribeToSelf
    }

    /**
     * subscribeToSelf property setter
     *
     * @param subscribeToSelf New subscribeToSelf value
     * @return this
     */
    fun setSubscribeToSelf(subscribeToSelf: Boolean): Room {
        this.subscribeToSelf = subscribeToSelf
        return this
    }

    /**
     * Runs all queued methods called while subscription was in progress
     */
    protected open fun dequeue() {
        if (this.queue.size > 0) {
            val threadPool = Executors.newSingleThreadExecutor()

            for (r in this.queue) {
                threadPool.execute(r)
            }

            threadPool.shutdown()

            try {
                threadPool.awaitTermination(1, TimeUnit.SECONDS)
            } catch (e: InterruptedException) {
                // do nothing
            } finally {
                this.queue.clear()
            }
        }
    }
}
/**
 * Constructor
 *
 * @param kuzzleDataCollection Data collection to link
 */
